### PR TITLE
write lcov coverage to a file - cmd line reporting

### LINKED ISSUE #13: Any way to run code coverage via command line?
I'm new to ember-cli and Ember in general, but is there a way to run blanket from command line? Any options to enable it via `ember test`? Moreover, is there an easy way to generate lcov reports to feed in to services like coveralls / code climate?

### PR DESCRIPTION
**Pull request by jschilli**
For discussion only: resolves #13

This sets up a testemMiddleware and installs a custom reporter for blanket

When the tests are complete we post the lcov data to the middleware and write to lcov.txt

```
SF:xx-utils/helpers/format-money
DA:1,1
DA:5,1
DA:7,1
DA:32,1
DA:33,0
DA:37,0
DA:38,0
DA:41,0
end_of_record
```
Open issues:

[] Biggest issue is that the lcov data is tied to the moduleName from app.js - when blanket runs in the browser it is annotating a cloned node from the original source in the browser.  This works for in-browser display as the line numbers match up.  We'll have to either reconcile the lcov.txt with the original source map files (e.g. postprocess) or evaluate other ways to match things up.
[] bunch of hardcoded stuff (url, output filename, etc)
[] figure out why `blanketReporter` (the mocha one) is being called twice

so not a bad start, but much to do

### CODE REVIEW

**Comment by jschilli** [comment]
[] either add `blanket-reporter.js` to blanket-loader.js or update blueprint to add to tests.html

I'll get back to this tonight or tomorrow - upon reflection I think doing the LCOV generation in browserland is wrong - I'll update the PR to serialize the coverageData directly from blanket to the middleware.

nodeland is a much friendlier place to do output formatting and will allow us to support many report formats (lcov, cobertura, etc) and deal with the sourcemap reconciliation.

I'm interested in any thoughts folks have about the sourcemap conundrum

**Comment by sglanzer** [comment]
This is a fantastic starting point for the discussion, thanks Jeff.

I definitely agree that LCOV generation should be driven from nodejs - I'm assuming phantomjs is probably the right place to start with this.  Have you had any luck with the headless running via ember test?

**Comment by jschilli** [comment]
Yes. Ember test writes Lcov.txt in the root directory. Updated PR coming tomorrow with a new reporting approach 

-jeff

**Comment by jschilli** [comment]
So good news and bad news:

Good: we are now writing `lcov.txt` to the root directory when running `ember test`
Bad: The output data suffers from an impedance mismatch

`ember-cli` transforms the data through `ember-cli-babel` (es6->es5) and then applies the es6 module transforms resulting in app.js and vendor.js

The blanket transformer (currently running in the browser) then transforms the loaded code, adding line and branch tracking statements.

So an input file (`app/utils/maths.js`)

starts thusly:

```javascript
var addNumbers,
    average,
    sumArray,
    averageArray,
    getAverage,
    percentDelta;

addNumbers = function(a, b) {
    var aIsNan = isNaN(a),
        bIsNan = isNaN(b);

    if (aIsNan && bIsNan) {
        return 0;
    }

```

In app.js

```javascript
define('xx-utils/utils/maths', ['exports'], function (exports) {
    'use strict';

    var addNumbers,
        average,
        sumArray,
        averageArray,
        getAverage,
        percentDelta;

    exports.addNumbers = addNumbers = function(a, b) {
        var aIsNan = isNaN(a),
            bIsNan = isNaN(b);

        if (aIsNan && bIsNan) {
            return 0;
        }
```

In the browser, the markedup source looks like:

```javascript
["define(\"xx-utils/utils/maths\", "
	, "  [\"exports\"], "
	, "function (exports) {", 
	"", 
	"    'use strict';", 
	"", 
	"    var addNumbers,",
	"        average,", 
	"        sumArray,", 
	"        averageArray,", 
	"        getAverage,", 
	"        percentDelta;", 
	"", 
	"    exports.addNumbers = addNumbers = function(a, b) {", 
	"        var aIsNan = isNaN(a),", 
	"            bIsNan = isNaN(b);", 
	"", "        if (aIsNan && bIsNan) {", 
	"            return 0;", "        }", 
	"        if (aIsNan) {", 
	"            return b;", 
```

The line numbers that are reported from blanket relate solely to the last file's structure, including the gratuitious linebreaks between `exports` and `function (exports)`

So no derivable coorelation between what we started from and the output

I've delved into the generated source maps using (https://github.com/mozilla/source-map/) but these get us only back to the intermediate form (where the file is located in app.js or vendor.js)

The testemMiddleware (https://github.com/sglanzer/ember-cli-blanket/pull/38/files#diff-168726dbe96b3ce427e7fedce31bb0bcR32) is providing the glue between the browser and ember-cli.

The browser fires off a `POST` (https://github.com/sglanzer/ember-cli-blanket/pull/38/files#diff-c1a0ca0ae9997d37108b9580a48466bcR59) with the results of the coverage run which the middleware then writes to the filesystem.  No need to modify phantom 

I'm going to leave this up for discussion and further exploration but IMO not mergeable - even with cleanup until we reconcile and can stand behind the reported data

**Comment by sglanzer** [comment]
Isn't this the case with all of the reported coverage so far?  The ES6 -> ES5 conversion makes a mess of things and the additional wrapping in the module definition ('define') doesn't help matters either.

I would absolutely love to fix the sourcemap issue (there was a mention of better sourcemaps being available in ember-cli a month or so back, not sure what the status is), but I wouldn't necessarily block the headless runner because of that - it's really a separate issue that the web client reporting also suffers from.

**Comment by jschilli** [comment]
The issue as I see it is less about line number matching between the covered code and the original (potentially es6) code than between the code that leaves ember-cli and that which is annotated by blanket.

So the real question is what do people do with the generated lcov/cobertura data? 

In my case i want to visualize the uncovered lines/branches.

secondarily I want an extractable value that I can use in a CI environment to graph coverage/prevent success if coverage < last value etc

So we have been chasing lcov output here but we have all the data that the web reports have - we could (as a first pass) generate an HTML report with the coverage data embedded - the coverage lines are self relative to the annotated code.

A different reporter could output JSON stats

We can emit LCOV data in LCOV format but it’s not useful IMO as raw data without being able to correctly visualize the covered/uncovered source lines

**Comment by sglanzer** [comment]
Let me summarize to see if I understand the conceptual issue:

- LCOV reports the same lines/coverage as the web based reporter currently
- This is reasonably acceptable in the web based reporter because it is paired with the converted ES6 code so that the corresponding lines are visible
- This is not acceptable in the LCOV report because there is no visible code pairing and the LCOV lines don't align to the original source

Is this accurate?

**Comment by jschilli** [comment]
Yes.

So now we have a mechanism to persist coverage data - what do we want to persist?

**Comment by sglanzer** [comment]
Cool, I definitely understand your hesitation then.

At a minimum we can currently report coverage percentages and the ideal goal would be to provide accurate mappings back to uncovered source code.  If you believe it's acceptable to move this in stages, then the percentage data extracted from the LCOV report could be paired with the web based report as a temporary measure.  Not the end goal, but workable in my opinion.

Basically - "if the automated LCOV measurements drop below 80%, run the web based LCOV report and look for under-covered sections of the code".

Is this a reasonable approach from your perspective?  I believe people would be excited just to have some data, even if it is just percentages, generated from a headless run.

**Comment by jschilli** [comment]
I’ll polish up the PR and add a `stats` reporter - reporters are setup to be pluggable - we can document that and see if anyone else riffs on it - and of course I’ll keep plugging away.

One other thing (and we can do this in a separate pr) is that `lib/` is becoming noisy with browserland and nodeland files comingled - I’d like to refactor that to have a `lib/client` dir to house the client destined files - we can do the magic in index.js to preserve the existing output/import into tests.html

I can have something in the next few days

**Comment by acorncom** [comment]
> I believe people would be excited just to have some data, even if it is just percentages, generated from a headless run

Yes indeed, I'm watching this repo because of just the possibility that you guys will be able to work out the LCOV stuff ;-)  I haven't been able to find anyone working on it right now, but it'd make a big difference for my team.  Thanks guys!

**Comment by sglanzer** [comment]
Sounds like a good start to me, thanks Jeff.

For the source mapping, I think we need to engage the ember-cli core team to see if there are any suggestions.

**Comment by jschilli** [comment]
sure 

cc: @rwjblue @steffanpenner

any thoughts you guys may have would be appreciated.

**Comment by jschilli** [comment]
also big caveat on current WIP PR - just realized that this is only working for mocha projects - I've annotated a qunit project and will push a sample project when I have it working as well

**Comment by jschilli** [comment]
I've pushed an update to #outputCoverage that includes qunit support

I've written a JSON reporter that outputs a file with stats

a sample is here: https://gist.github.com/jschilli/c720f73b9257274c8ef3

A project with working command line coverage is here: 

https://github.com/jschilli/todo-mvc-ember-cli-tdd/tree/qunit-blanket

Seemed like a reasonable project to test coverage on

The blanket-options.js file format changed to support consumption by both browsers and nodeland

```javascript
/* globals blanket */
var options = {
    modulePrefix: "todomvc-ember-cli",
    filter: "//.*todomvc-ember-cli/.*/",
    antifilter: "//.*(tests|template).*/",
    modulePattern: "\/(\\w+)",
    branchTracking: true,
    loaderExclusions: [],
    enableCoverage: true,
    cliOptions: {
      outputFile: 'test-output.json',
      reporters: ['json']
    }
  };

if (typeof exports === undefined) {
  blanket.options(options);
} else {
  module.exports = options;
}

```

(the cool discovery here was modulePattern which summarizes stats by module - supported in the `json` reporter)

There is still some cleanup and testing to do but feel free to give it a whirl.

I've also started a test suite `npm test` for the project

I open for feedback - this is still WIP but hope to wrap something up over the weekend or early next week

**Comment by jschilli** [comment]
I think this one is ready to go.

Would love eyes on some of the decisions.

Tested and working with QUnit and Mocha

test qunit project at: https://github.com/jschilli/todo-mvc-ember-cli-tdd/tree/qunit-blanket

File: blueprints/ember-cli-blanket/index.js [hunk 1]
```diff
@@ -35,31 +33,6 @@ module.exports = {
               '    <style>#blanket-main { position: relative; z-index: 99999; }</style>',
               { after: '<link rel="stylesheet" href="assets/test-support.css">' + EOL }
           );
-        }.bind(this))
-
-        .then(function() {
-            var fullPath = path.join(this.project.root, 'tests/blanket-options.js');
```
Comment (jschilli) [inline blueprints/ember-cli-blanket/index.js]: I moved all of this into a blueprint file.

Updating by rerunning `ember g ember-cli-blanket` gives you the familiar Overwrite blanket-options.js [Yndeh] prompt

File: index.js [hunk 2]
```diff
@@ -1,76 +1,106 @@
...
+var coverageMiddleware = require('./lib/coverage-middleware');
+var bodyParser = require('body-parser');
+
+function logErrors(err, req, res, next) {
+  console.error(err.stack);
+  next(err);
+}
+
 module.exports = {
-    name: 'Ember CLI Blanket',
```
Comment (jschilli) [inline index.js]: reformatted the whole file - we've got 4 spaces and 2 spaces etc, etc - needs a cleanup overall

File: index.js [hunk 3]
```diff
@@ -1,76 +1,106 @@
...
+
+      var vendor = this.treeGenerator(path.join(__dirname, 'vendor'));
+
+      var blanketLoader = new Funnel(vendor, {
+        files: ['blanket-reporter.js', 'blanket-require.js'],
+        srcDir: '/',
+        destDir: '/'
+      });
+
+      blanketLoader = this.concatFiles(blanketLoader, {
```
Comment (jschilli) [inline index.js]: we're concatting multiple needed files into one - the rename on blanket-loader.js to blanket-require.js was to reduce confusion by Chrome on sourcemaps - source file and concatted file were the same name

File: lib/blanket-options.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+var fs = require('fs-extra');
+
+function exists(path) {
+  return fs.existsSync(path);
+}
+
+function isCurrentFormat(path) {
```
Comment (jschilli) [inline lib/blanket-options.js]: This allows us to detect an upgrade needed for the blanket-options.js file

File: lib/coverage-middleware.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  return blanketOptions;
+}
+
+function reporterForName(reporterName) {
+  return require(path.join(__dirname, 'reporters', reporterName + '-reporter'));
+}
+
+// TODO handle multiple reporters
+function loadReporter(blanketOptions) {
+  var reporterName = blanketOptions.cliOptions.reporters[0] || 'null';
```
Comment (jschilli) [inline lib/coverage-middleware.js]: Use the first reporter listed or when none are present the `null` or `no-op` reporter

Comment (jschilli) [inline lib/coverage-middleware.js]: this file needs tests

File: vendor/blanket-reporter.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/*global blanket */
```
Comment (jschilli) [inline vendor/blanket-reporter.js]: Moved the client side files to vendor/ to avoid confusion

File: blueprints/ember-cli-blanket/files/tests/blanket-options.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/*globals blanket */
+
+var options = {
+  modulePrefix: "<%= dasherizedPackageName %>",
```
Comment (jschilli) [inline blueprints/ember-cli-blanket/files/tests/blanket-options.js]: This is the new blueprint file for the blanket-options.js generation

File: lib/blanket-options.js (hunk 4 above)
Comment (sglanzer) [inline lib/blanket-options.js]: I'm assuming modifications to blanket-options will be normal.  As noted above this will only result in a [Yndeh] prompt on upgrade if the contents don't match, correct?

File: lib/reporters/lcov-reporter.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+
+var Reporter = require('../reporter');
+
+var lcovRecord = function(data) {
+
+  var str = "";
+  str += 'SF:' + data.fileName + '\n';
```
Comment (sglanzer) [inline lib/reporters/lcov-reporter.js]: What do 'SF:' and 'DA:' stand for?  I understand the data your putting here, just not the labelling.

File: package.json [hunk 9]
```diff
@@ -33,23 +34,32 @@
...
+    "body-parser": "^1.12.2",
     "broccoli-funnel": "^0.2.2",
-    "broccoli-merge-trees": "^0.2.1"
+    "broccoli-merge-trees": "^0.2.1",
+    "fs-extra": "^0.18.0",
+    "core-object": "^1.1.0",
+    "lodash": "^3.6.0"
   },
   "ember-addon": {
     "after": "ember-cli-qunit",
```
Comment (sglanzer) [inline package.json]: Does this apply to mocha as well?  Do we need this anymore?

**Comment by sglanzer** [comment]
Wow, this is a huge amount of work, fantastic job Jeff!

File: lib/reporters/lcov-reporter.js (hunk 8 above)
Comment (jschilli) [inline lib/reporters/lcov-reporter.js]: See http://ltp.sourceforge.net/coverage/lcov/geninfo.1.php

There are more data markers that we can add - this is the basics.

File: package.json (hunk 9 above)
Comment (jschilli) [inline package.json]: Fair question on the after hook. I'll give it a shot when I take a final pass on this

Plan is to squash (will be Monday before I can do this)

File: blueprints/ember-cli-blanket/files/tests/blanket-options.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+/*globals blanket, module */
+
+var options = {
+  modulePrefix: "<%= dasherizedPackageName %>",
+  filter: "//.*<%= dasherizedPackageName %>/.*/",
+  antifilter: "//.*(tests|template).*/",
+  loaderExclusions: [],
+  enableCoverage: true,
+  cliOptions: {
+    outputFile: 'test-output.json',
```
Comment (jschilli) [inline blueprints/ember-cli-blanket/files/tests/blanket-options.js]: I'm going to create reporter specific options so output files would be under `jsonOptions:{}`

File: lib/reporters/lcov-reporter.js (hunk 8 above)
Comment (sglanzer) [inline lib/reporters/lcov-reporter.js]: Thanks! This is my first foray into lcov, that's useful info.  Adding that link as a comment would be helpful for any future work.

**Comment by tim-evans** [comment]
Yess. Having this for TravisCI will be amazing.

**Comment by amkirwan** [comment]
This would be great to have for ci testing.

**Comment by jschilli** [comment]
I think it’s good to go - will merge in the next day or so and then we’ll cut a release

**Comment by jschilli** [comment]
@sglanzer can you do the needful to publish a release?

I'll close the impacted tickets and open up new issues for the additional work (multiple reporters, additional reporters etc)

**Comment by sglanzer** [comment]
Absolutely!
